docs/gcli.1
.Dd $Mdocdate$
.Dt GCLI 1
.Os
.Sh NAME
.Nm gcli
.Nd interact with git forges without using a web-browser
.Sh SYNOPSIS
.Pp
.Nm
.Op overrides
.Cm subcommand
.Op options
.Sh DESCRIPTION
.Nm
can be used to interact with
.Xr git 1
forges like GitHub, GitLab and Gitea from the command line in order to
make many tasks like managing issues and pull requests easier.
.Pp
Calls to
.Nm
usually consist of either only the subcommand to list requested data
or the subcommand plus further subcommands or options to perform
various tasks. Some commands may also take an item to operate on and
accept multiple actions that will be performed on the item (e.g. PRs
may be summarized, comments fetched and a diff printed all in one
command).
.Pp
The default behaviour of
.Nm
can be overriden to accomodate more nuanced use cases. Manual
overrides must be passed before subcommands and their options.
.Sh SUBCOMMANDS
Most of these subcommands are documented in dedicated man pages.
.Bl -tag -width milestones
.It Cm issues
Issues in repositories. See
.Xr gcli-issues 1 .
.It Cm pulls
Pull Requests on repositories. See
.Xr gcli-pulls 1 .
.It Cm labels
Manage labels for issues and pull/merge requests on repositories. See
.Xr gcli-labels 1 .
.It Cm forks
Forking repositories. See
.Xr gcli-forks 1 .
.It Cm gists
Github Gists are like paste bins to where you can dump code snippets
etc. See
.Xr gcli-gists 1 .
.It Cm snippets
Support for Gitlab snippets. See
.Xr gcli-snippets 1 .
.It Cm repos
Manage your own or other repositories. See
.Xr gcli-repos 1 .
.It Cm comment
Submit comments under issues and PRs. See
.Xr gcli-comment 1 .
.It Cm status
Print a list of TODOs and/or notifications. See
.Xr gcli-status 1 .
.It Cm pipelines
Inspect and manage Gitlab Pipelines. See
.Xr gcli-pipelines 1 .
.It Cm releases
Create and manage releases. See
.Xr gcli-releases 1 .
.It Cm milestones
List and manage milestones. See
.Xr gcli-milestones 1 .
.It Cm api
Perform direct queries to the API and dump the JSON response to
stdout. This is primarily intended to assist debugging gcli. See
.Xr gcli-api 1 .
.It Cm version
Print version and exit.
.El
.Sh OPTIONS
.Nm
overrides are:
.Bl -tag -width indent
.It Fl a , -account Ar override-account
Manually override the default account.
.Ar override-account
must name a config section for an account in the global config file. See
.Sx FILES .
.It Fl r , -remote Ar override-remote
Use
.Ar override-remote
as the remote when trying to infer repository data.
.It Fl c , -colours
Ignore
.Ev NO_COLOR
as well as whether the output is not tty and print ANSI escape
sequences for changing text formatting. Default is to output colours
unless stdout is not a tty. See
.Xr isatty 3 .
This is useful in combination with modern pagers such as
.Xr less 1 .
.It Fl q , -quiet
Supresses most output of
.Nm .
.It Fl v , -verbose
Be very verbose. This means that warnings about missing config files
and request steps are printed to stderr.
.It Fl t , -type Ar forge-type
Forcefully override the forge type. Set
.Ar forge-type
to
.Sq github ,
.Sq gitlab
or
.Sq gitea
to connect to the corresponding services.
.El
.Pp
Common options across almost all of the subcommands are:
.Bl -tag -width indent
.It Fl s , -sorted
Reverse the output such that most recent items appear at the bottom.
.It Fl n , -count Ar n
Fetch multiple pages of data. The default is usually 30 items, but
this parameter allows to fetch more than that. Setting
.Ar n
to -1 will result in all pages being queried and all items being read.
However, be careful with that, since if there is a lot of data to be
fetched, it may result in rate limiting by the Github API, aside from
the fact that it may also take a considerable amount of time to
process.
.It Fl a , -all
Fetch all data, including closed issues and closed/merged PRs.
.It Fl y , -yes
Do not ask for confirmation when performing destructive operations or
performing submissions. Always assume yes.
.It Fl o , -owner Ar owner
Operate on the given owner (organization or user).  Can only be used
in combination with
.Fl r .
.It Fl r , -repo Ar repo
Operate on the given repository.  Can only be used in combination with
.Fl o .
.It Fl i Ar id
Operate on the given numeric identifier.
.El
.Pp
Other options specific to the context are documented in the respective
man pages.
.\" .Sh IMPLEMENTATION NOTES
.\" Not used in OpenBSD.
.Sh ENVIRONMENT
.Bl -tag -width XDG_CONFIG_DIR
.It Ev EDITOR
If the gcli config file does not name an editor,
.Nm
may use this editor.
.It Ev XDG_CONFIG_DIR
There should be a subdirectory called gcli in the directory this
environment variable points to where
.Nm
will go looking for its configuration file. See
.Sx FILES .
.It Ev GCLI_ACCOUNT
Specifies an account name that should be used instead of an inferred
one. The value of
.Ev GCLI_ACCOUNT
can be overridden again by using
.Fl a Ar account-name .
This is helpful in cases where you have multiple accounts of the same
forge-type configured and you don't want to use the default.
.It Ev NO_COLOR
If set to
.Sq 1 ,
.Sq y or
.Sq yes
(capitalization ignored) this will suppress output of ANSI colour
escape sequences. See
.Sx OPTIONS
(--colours).
.El
.Sh FILES
.Bl -tag -width ${XDG_CONFIG_DIR}/gcli/config -compact
.It Pa ${XDG_CONFIG_DIR}/gcli/config
The config file for
.Nm .
It shall contain the following data:
.Pp
.Bd -literal
defaults {
	editor=/path/to/ganoooo/emacs
	github-default-account=herrhotzenplotz-gh
	gitlab-default-account=herrhotzenplotz-gitlab
}

herrhotzenplotz-gh {
	account=herrhotzenplotz
	token=foobar
	apibase=https://api.github.com
	forge-type=github
}

herrhotzenplotz-gl {
	account=herrhotzenplotz
	token=<valid gitlab api token>
	apibase=https://gitlab.com/api/v4
	forge-type=gitlab
}
.Ed
.Pp
In case
.Sq apibase
is not set, it defaults to the above values.
For the API token, you can set whatever scopes you want. However, I
recommend setting the following on GitHub:
.Sq admin:org, delete_repo, gist, repo, workflow .
On GitLab you only need the
.Sq api
scope.
.Pp
If editor is not set in the config file,
.Nm
will use
.Ev EDITOR
from the environment.
.Pp
Both
.Sq gitlab-default-account
and
.Sq github-default-account
must point at a config section with that exact name.
.Pp
.It Pa .gcli
A repo-specific config file that may contain the following data:
.Bd -literal
pr.upstream=herrhotzenplotz/gcli
pr.base=trunk
.Ed
.Pp
It is intended to be committed into the repo so that users don't have
to manually specify all the options like
.Fl -in ,
.Fl -from ,
.Fl -base etc.
.Pp
.El
.Sh EXAMPLES
List recently opened issues in the current upstream repository:
.Bd -literal -width indent
$ gcli issues
.Ed
.Pp
Merge upstream PR #22:
.Bd -literal -width indent
$ gcli pulls -p 22 merge
.Ed
.Pp
Get a summary and comments of upstream PR #22:
.Bd -literal -width indent
$ gcli pulls -p 22 summary comments
.Ed
.Pp
Establish a connection to github and print the last 10 pull requests
in contour-terminal/contour regardless of their state.
.Bd -literal -width indent
$ gcli -t github pulls -o contour-terminal -r contour -a -n10
.Ed
.Pp
This can be useful if neither your config file nor the directory you're
working from contain the relevant forge and repository information.
.Sh SEE ALSO
.Xr git 1 ,
.Xr gcli-issues 1 ,
.Xr gcli-pulls 1 ,
.Xr gcli-labels 1 ,
.Xr gcli-comment 1 ,
.Xr gcli-review 1 ,
.Xr gcli-forks 1 ,
.Xr gcli-repos 1 ,
.Xr gcli-gists 1 ,
.Xr gcli-releases 1 ,
.Xr gcli-comment 1
.Xr gcli-pipelines 1
.\" .Sh STANDARDS
.Sh HISTORY
The idea for
.Nm
appeared during a long rant on IRC where the issue with the official
tool written by GitHub became clear to be the manual dialing and DNS
resolving by the Go runtime, circumventing almost the entirety of the
IP and DNS services of the operating system and leaking sensitive
information when using Tor.
.Pp
Implementation started in October 2021 with the goal of having a
decent, sufficiently portable and secure version of a cli utility to
interact with the GitHub world without using the inconvenient web
interface.
.Pp
Later, support for GitLab and Gitea (Codeberg) were added.
.Sh AUTHORS
.An Nico Sonack aka. herrhotzenplotz Aq Mt nsonack@herrhotzenplotz.de
.Sh CAVEATS
Not all features that are available from the web version are available in
.Nm .
However, it is a non-goal of the project to provide all this
functionality.
.Sh BUGS
Yes. It is software.
.Pp
Please report issues preferably via e-mail, on GitLab or on
GitHub. You may also report an issue like so:
.Bd -literal -width indent
$ gcli -a some-gitlab-account issues create -o herrhotzenplotz -r gcli "BUG : ..."
.Ed
.Sh SECURITY CONSIDERATIONS
It is written in C. If it were written in Rust, it would have been
much safer.


src/gcli.c
/*
 * Copyright 2021,2022 Nico Sonack <nsonack@herrhotzenplotz.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following
 * disclaimer in the documentation and/or other materials provided
 * with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config.h>

#ifdef HAVE_GETOPT_h
#include <getopt.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <gcli/cmd.h>
#include <gcli/config.h>

static void usage(void);

static int
subcommand_version(int argc, char *argv[])
{
	(void) argc;
	(void) argv;

	version();
	copyright();

	return EXIT_SUCCESS;
}

static struct subcommand {
	char const *const cmd_name;
	char const *const docstring;
	int (*fn)(int, char **);
} subcommands[] = {
	{ .cmd_name = "ci",
	  .fn = subcommand_ci,
	  .docstring = "Github CI status info" },
	{ .cmd_name = "comment",
	  .fn = subcommand_comment,
	  .docstring = "Comment under issues and PRs" },
	{ .cmd_name = "forks",
	  .fn = subcommand_forks,
	  .docstring = "Create, delete and list repository forks" },
	{ .cmd_name = "gists",
	  .fn = subcommand_gists,
	  .docstring = "Create, fetch and list Github Gists" },
	{ .cmd_name = "issues",
	  .fn = subcommand_issues,
	  .docstring = "Manage issues" },
	{ .cmd_name = "labels",
	  .fn = subcommand_labels,
	  .docstring = "Manage issue and PR labels" },
	{ .cmd_name = "milestones",
	  .fn = subcommand_milestones,
	  .docstring = "Milestone handling" },
	{ .cmd_name = "pipelines",
	  .fn = subcommand_pipelines,
	  .docstring = "Gitlab CI management" },
	{ .cmd_name = "pulls",
	  .fn = subcommand_pulls,
	  .docstring = "Create, view and manage PRs" },
	{ .cmd_name = "releases",
	  .fn = subcommand_releases,
	  .docstring = "Manage releases of repositories" },
	{ .cmd_name = "repos",
	  .fn = subcommand_repos,
	  .docstring = "Remote Repository management" },
	{ .cmd_name = "snippets",
	  .fn = subcommand_snippets,
	  .docstring = "Fetch and list Gitlab snippets" },
	{ .cmd_name = "status",
	  .fn = subcommand_status,
	  .docstring = "General user status and notifications" },
	{ .cmd_name = "api",
	  .fn = subcommand_api,
	  .docstring = "Fetch plain JSON info from an API (for debugging purposes)" },
	{ .cmd_name = "version",
	  .fn = subcommand_version,
	  .docstring = "Print version" },
};

static void
usage(void)
{
	fprintf(stderr, "usage: gcli [options] subcommand\n\n");
	fprintf(stderr, "OPTIONS:\n");
	fprintf(stderr, "  -a account     Use the configured account instead of inferring it\n");
	fprintf(stderr, "  -r remote      Infer account from the given git remote\n");
	fprintf(stderr, "  -t type        Force the account type:\n");
	fprintf(stderr, "                    - github (default: github.com)\n");
	fprintf(stderr, "                    - gitlab (default: gitlab.com)\n");
	fprintf(stderr, "                    - gitea (default: codeberg.org)\n");
	fprintf(stderr, "  -c             Force colour and text formatting.\n");
	fprintf(stderr, "  -q             Be quiet. (Not implemented yet)\n\n");
	fprintf(stderr, "  -v             Be verbose.\n\n");
	fprintf(stderr, "SUBCOMMANDS:\n");
	for (size_t i = 0; i < ARRAY_SIZE(subcommands); ++i) {
		fprintf(stderr,
		        "  %-13.13s  %s\n",
		        subcommands[i].cmd_name,
		        subcommands[i].docstring);
	}
	fprintf(stderr, "\n");
	version();
	copyright();
}

int
main(int argc, char *argv[])
{
	/* Parse first arguments */
	if (gcli_config_init(&argc, &argv)) {
		usage();
		return EXIT_FAILURE;
	}

	/* Make sure we have a subcommand */
	if (argc == 0) {
		fprintf(stderr, "error: missing subcommand\n");
		usage();
		return EXIT_FAILURE;
	}

	/* Find and invoke the subcommand handler */
	for (size_t i = 0; i < ARRAY_SIZE(subcommands); ++i) {
		if (strcmp(subcommands[i].cmd_name, argv[0]) == 0)
			return subcommands[i].fn(argc, argv);
	}

	/* No subcommand matched */
	fprintf(stderr, "error: unknown subcommand %s\n", argv[0]);
	usage();

	return EXIT_FAILURE;
}
